### PR TITLE
Change "synchronized" to reentrant lock for virtual-threads

### PR DESCRIPTION
**Pull request by omercelikceng**
This commit ensures that the method is friendly for virtual threads to avoid blocking and pinning. The lock is acquired at the beginning of the method and released in a finally block to ensure it is always released, even if an exception occurs.

### CODE REVIEW

File: spring-kafka/src/main/java/org/springframework/kafka/core/KafkaAdmin.java [hunk 1]
```diff
@@ -267,12 +272,16 @@ public final boolean initialize() {
 			}
 			if (adminClient != null) {
 				try {
-					synchronized (this) {
+					try {
```
Comment (sobychacko) [inline spring-kafka/src/main/java/org/springframework/kafka/core/KafkaAdmin.java:275]: Can we use only one `try` block? I think, you can move the unlocking to the `finally` block on the main `try`.

File: spring-kafka/src/main/java/org/springframework/kafka/support/serializer/JsonSerializer.java [hunk 2]
```diff
@@ -147,30 +152,36 @@ public void setUseTypeMapperForKey(boolean isKey) {
...
-		}
-		Assert.state(!this.setterCalled
-				|| (!configs.containsKey(ADD_TYPE_INFO_HEADERS) && !configs.containsKey(TYPE_MAPPINGS)),
-				"JsonSerializer must be configured with property setters, or via configuration properties; not both");
-		setUseTypeMapperForKey(isKey);
-		if (configs.containsKey(ADD_TYPE_INFO_HEADERS)) {
-			Object config = configs.get(ADD_TYPE_INFO_HEADERS);
-			if (config instanceof Boolean configBoolean) {
-				this.addTypeInfo = configBoolean;
+		try {
```
Comment (sobychacko) [inline spring-kafka/src/main/java/org/springframework/kafka/support/serializer/JsonSerializer.java:155]: Forgot to remove the `synchronized` keyword from the method signature?

**Review by sobychacko — CHANGES_REQUESTED** [review]
See my in-line comments.

File: spring-kafka/src/main/java/org/springframework/kafka/core/KafkaAdmin.java (hunk 1 above)
Comment (omercelikceng) [reply spring-kafka/src/main/java/org/springframework/kafka/core/KafkaAdmin.java:275]: Hi @sobychacko .
If I do what you suggest, I will end up locking the method, including the addOrModifyTopicsIfNeeded method. I reviewed the addOrModifyTopicsIfNeeded method, and I think it contains many operations. In my opinion, this method should not be included within the lock.

Comment (sobychacko) [reply spring-kafka/src/main/java/org/springframework/kafka/core/KafkaAdmin.java:275]: That makes sense. I guess you can try to extract that inner try block into a separate private method - something like `updateClusterId(...)` and then call that in the current method. That way, we can avoid stacking one `try` block on another and thus look less cluttered.

Comment (omercelikceng) [reply spring-kafka/src/main/java/org/springframework/kafka/core/KafkaAdmin.java]: You're right, I'll fix it right away.

**Comment by omercelikceng** [comment]
> See my in-line comments.

@sobychacko I've made the updates; could you review them again?

**Comment by sobychacko** [comment]
@omercelikceng Thanks for the PR. Merged if via https://github.com/spring-projects/spring-kafka/commit/3c97a3bce1ed8141f7b5ff8390c0a68f6f1e232b.

**Pull request closed without merging**